test-results.cake

```diff
@@ -0,0 +1,182 @@
+    private string GetAttribute(XmlNode node, string name)
+    {
+        return node.Attributes[name]?.Value;
+    }
+
+    private int IntAttribute(XmlNode node, string name)
+    {
+        string s = GetAttribute(node, name);
+        return s == null ? 0 : int.Parse(s);
+    }
+}
+
+public class TestReport
+{
+    public PackageTest Test;
+    public ActualResult Result;
+    public List<string> Errors;
+
+    public TestReport(PackageTest test, ActualResult result)
+    {
+        Test = test;
+        Result = result;
+        Errors = new List<string>();
+
+        var expected = test.ExpectedResult;
+
+        ReportMissingFiles();
+
+        if (result.OverallResult == null)
+            Errors.Add("     The test-run element has no result attribute.");
+        else if (expected.OverallResult != result.OverallResult)
+            Errors.Add($"     Expected: Overall Result = {expected.OverallResult}\n    But was: {result.OverallResult}");
+        CheckCounter("Test Count", expected.Total, result.Total);
+        CheckCounter("Passed", expected.Passed, result.Passed);
+        CheckCounter("Failed", expected.Failed, result.Failed);
+        CheckCounter("Warnings", expected.Warnings, result.Warnings);
+        CheckCounter("Inconclusive", expected.Inconclusive, result.Inconclusive);
+        CheckCounter("Skipped", expected.Skipped, result.Skipped);
+    }
+
+    public TestReport(PackageTest test, Exception ex)
+    {
+        Test = test;
+        Result = null;
+        Errors = new List<string>();
+        Errors.Add($"     {ex.Message}");
+    }
+
+    public void Display(int index)
+    {
+        Console.WriteLine($"\n{index}. {Test.Description}");
+        Console.WriteLine($"   Args: {Test.Arguments}\n");
+
+        foreach (var error in Errors)
+            Console.WriteLine(error);
+
+        Console.WriteLine(Errors.Count == 0
+            ? "   SUCCESS: Test Result matches expected result!"
+            : "\n   ERROR: Test Result not as expected!");
+    }
+
+    private void ReportMissingFiles()
+    {
+        var suites = Result.Xml.SelectNodes(
+            "//test-suite[@type='Unknown'] | //test-suite[@type='Project'] | //test-suite[@type='Assembly']");
+        if (suites.Count == 0)
+            Errors.Add("     No top-level suites! Possible empty command-line or misformed project.");
+
+        foreach (XmlNode suite in suites)
+        {
+            string suiteResult = GetAttribute(suite, "result");
+            string label = GetAttribute(suite, "label");
+            string site = suite.Attributes["site"]?.Value ?? "Test";
+            if (suiteResult == "Failed" && site == "Test" && label == "Invalid")
+            {
+                string message = suite.SelectSingleNode("reason/message")?.InnerText;
+                Errors.Add($"     {message}");
+            }
+        }
+    }
+
+    private void CheckCounter(string label, int expected, int actual)
+    {
+        if (expected > 0 && expected != actual)
```

Comment:
I think this ties into my comment above on zero being a valid value here, I'm not sure expected needs to be > 0, does it?

Reply:
Slightly different here. However, it should be >= 0 because ExpectedValue uses -1 to mean "this doesn't need to be checked." If we wanted to run one of our __actual__ tests as a package test, I'd just require it to pass and wouldn't check all the counts because it would require too much maintenance.